File: GLLara/GLLVertexAttribAccessorSet.m
//
//  GLLVertexAttribAccesorSet.m
//  GLLara
//
//  Created by Torsten Kammer on 07.04.21.
//  Copyright © 2021 Torsten Kammer. All rights reserved.
//

#import "GLLVertexAttribAccessorSet.h"

#import "GLLVertexAttribAccessor.h"
#import "GLLVertexFormat.h"

#import "NSArray+Map.h"

@implementation GLLVertexAttribAccessorSet

- (instancetype)initWithAccessors:(NSArray<GLLVertexAttribAccessor *> *)accessors {
    if (!(self = [super init]))
        return nil;
    
    _accessors = [accessors copy];
    
    return self;
}

- (GLLVertexAttribAccessorSet *)setByCombiningWith:(GLLVertexAttribAccessorSet *)other;
{
    return [[[self class] alloc] initWithAccessors:[self.accessors arrayByAddingObjectsFromArray:other.accessors]];
}

- (GLLVertexAttribAccessor *__nullable)accessorForSemantic:(enum GLLVertexAttribSemantic)semantic layer:(NSUInteger)layer {
    return [self.accessors firstObjectMatching:^BOOL(GLLVertexAttribAccessor *accessor) {
        return accessor.attribute.semantic == semantic && accessor.attribute.layer == layer;
    }];
}

- (GLLVertexAttribAccessor *__nullable)accessorForSemantic:(enum GLLVertexAttribSemantic)semantic {
    return [self accessorForSemantic:semantic layer:0];
}

- (GLLVertexFormat *)vertexFormatWithVertexCount:(NSUInteger)count hasIndices:(BOOL)hasIndices {
    return [[GLLVertexFormat alloc] initWithAttributes:[self.accessors map:^GLLVertexAttrib* (GLLVertexAttribAccessor *accessor) {
        return accessor.attribute;
    }] countOfVertices:count hasIndices:hasIndices];
}

@end

File: GLLara/GLLVertexAttribAccessorSet.swift
//
//  GLLVertexAttribAccessorSet.swift
//  GLLara
//
//  Created by Torsten Kammer on 06.04.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation
import Metal

@objc class GLLVertexAttribAccessorSet: NSObject {
    var accessors: [GLLVertexAttribAccessor]
    
    let vertexDescriptor: MTLVertexDescriptor
    
    @objc init(accessors: [GLLVertexAttribAccessor]) {
        self.accessors = accessors
        
        // Set up the vertex descriptor
        // - Keep track of which buffers are found in the layouts section
        var orderedBuffers: [Data] = []
        vertexDescriptor = MTLVertexDescriptor()
        vertexDescriptor.layouts[10].stepFunction = .perVertex
        vertexDescriptor.layouts[10].stride = accessors.first?.stride ?? 0
        for accessor in accessors {
            let index = accessor.attribute.identifier
            vertexDescriptor.attributes[index].offset = accessor.dataOffset
            vertexDescriptor.attributes[index].format = accessor.attribute.format
            vertexDescriptor.attributes[index].bufferIndex = 10
            
            if let data = accessor.dataBuffer {
                if let bufferIndex = orderedBuffers.firstIndex(of: data) {
                    vertexDescriptor.attributes[index].bufferIndex = bufferIndex
                } else {
                    let bufferIndex = orderedBuffers.count + 10
                    vertexDescriptor.attributes[index].bufferIndex = bufferIndex
                    orderedBuffers.append(data)
                    vertexDescriptor.layouts[bufferIndex].stepFunction = .perVertex
                    vertexDescriptor.layouts[bufferIndex].stride = accessor.stride
                }
            }
        }
    }
    
    func combining(with other: GLLVertexAttribAccessorSet) -> GLLVertexAttribAccessorSet {
        var adjustedAccessors = accessors
        adjustedAccessors.removeAll {
            other.accessor(semantic: $0.attribute.semantic, layer: $0.attribute.layer) != nil
        }
        return GLLVertexAttribAccessorSet(accessors: adjustedAccessors + other.accessors)
    }
    
    func accessor(semantic: GLLVertexAttribSemantic, layer: Int = 0) -> GLLVertexAttribAccessor? {
        return accessors.first {
            $0.attribute.semantic == semantic && $0.attribute.layer == layer
        }
    }
    
    func vertexFormat(vertexCount: Int, hasIndices: Bool) -> GLLVertexFormat {
        return GLLVertexFormat(attributes: accessors.map { $0.attribute }, countOfVertices: vertexCount, hasIndices: hasIndices)
    }
}
